index.bs

```diff
@@ -1641,6 +1645,28 @@ session.SystemProxyConfiguration = (
 
 </pre>
 
+#### The session.UserPromptHandler Type #### {#type-session-UserPromptHandler}
+
+[=Remote end definition=] and [=local end definition=]
+
+<pre class="cddl remote-cddl local-cddl">
+session.UserPromptHandler = session.UserPromptHandlerType /
+                            session.UserPromptHandlerConfiguration;
+
+session.UserPromptHandlerType = "accept" / "dismiss" / "ignore";
+
+session.UserPromptHandlerConfiguration = {
+  ? alert: session.UserPromptHandlerType,
+  ? beforeUnload: session.UserPromptHandlerType,
+  ? confirm: session.UserPromptHandlerType,
+  ? default: session.UserPromptHandlerType,
```

Comment:
If we want to extend it to file dialogs and auth prompts (it was mentioned previously), we probably should not have a default handler and require defining prompt handlers per type since we won't be able to support 'accept' for file and auth prompts.

Reply:
We could remove it from BiDi, but we'd still need to specify it for classic because historically that's only had a default handler, not per-type handlers. 

For `alert` `accept` and `dismiss` already do the same thing, so one option for `file` would be to make both of them act like `dismiss`, or make `accept` act like `ignore`. For HTTP Auth arguably `accept` and `dismiss` both make sense, although `accept` sending empty credentials does seem like it has a high chance of ending up in a loop.

Reply:
Perhaps it makes sense to only allow `dismiss` and `ignore` (as though seem to be supported by any prompt type) for the default UserPromptHandlerType in BiDi? I think it is perhaps not very important but my concern is that we will eventually be defining 'accept'  handlers where it makes no much sense just to be compatible with the list of possible default values.

Reply:
I think we'll need to define them anyway for as long as we still care about classic, because classic implementations previously only allowed a single string value as the prompt handler. So if we start invoking these algorithms for new prompt types (as we have for `beforeUnload`) we need to specify what happens with those existing string values. Passing an object with only the `default` key set is defined to be exactly the same as defining a single string value.

For `beforeUnload` the current decision is that the `default` value is ignored: if you don't provide an explicit `beforeUnload` key then the prompt is auto-dismissed. That matches the default behavior of classic which is to always close `beforeUnload` prompts.

By the same token we could ignore `default` for new prompt types and always use `ignore` unless a prompt-specific default is supplied. That would also match the legacy behaviour where we just don't handle these other modals at all.

On the other hand, in the long run, this approach of making `default` be "what would classic do?" is itself quite confusing for users or clients that don't care about the classic behaviour. 

So, just thinking about BiDi, we have several options:
* We can continue to support `default` and single string handler values, aligning the behaviour with whatever classic does in those cases. This has the disadvantage of being confusing e.g. it is unclear why `beforeUnload` doesn't respect `default`.
* We can have different semantics for `default` between classic and BiDi. For example we could make `beforeUnload` actually respect `default` for BiDi only sessions. This has the disadvantage of adding even more complexity in transitioning between classic and BiDi, and more complexity in the implementations, plus it doesn't really solve the question of how to handle `"default": "accept"` in the `file` case.
* We could subset the values allowed as default to just `dismiss` and `ignore`, assuming that they make sense for every prompt type, and force writing out any prompts you want to accept longhand e.g. `{"default": "dismiss", "confirm": "accept", "beforeUnload": "accept"}`. We'd also need a default if `default` isn't provided, which basically has the same problem that the legacy default for `alert`, `confirm` and `prompt` is `dismiss`, but `beforeUnload` is `accept` and `file` and `httpAuth` are effectively `ignore`.
* We could remove `default` (and the single-string values) altogether and force writing out everything longhand. That means something like `{"alert": "accept", "prompt": "accept", "confirm": "accept", "beforeUnload": "accept"}`. This is basically like the option above except the "what if default isn't specified" question only occurs when you don't specify a value at all (which is unfortunately probably the most common case).
* We could find some way to rename `default` that better specifies that it doesn't apply to all prompt types (but that doesn't help with string-only values).

I think if we were only concerned with BiDi we probably wouldn't have a default value, but explain that if no prompt handler is supplied the behaviour depends on the modal type, and that the keys override the default for that prompt type.

With the fact that classic has previously _only_ had the default override, and it didn't apply to all modals, it's less clear to me what the right option is.

Reply:
I would prefer this option:

> We can have different semantics for default between classic and BiDi. For example we could make beforeUnload actually respect default for BiDi only sessions. 

combined with

> We could subset the values allowed as default to just dismiss and ignore

But, in my opinion, the current proposal also looks good although I would find it confusing if we have eventually add more prompts that do not have the auto-accept behavior. Perhaps we could extend "accept" with some with extra parameters, for example, `{"alertParams": {"userText": "default value for accept"}, "fileParams": {"file": "/dev/null"}}`. Perhaps it also would be an option to allow prompt handler to be either `ignore` or an instance of `HandleUserPromptParameters` but I am not sure what the implications for that would be w.r.t to the classic compatibility.

So I think the current proposal looks good for now but I want @sadym-chromium to review this as well.

Reply:
"Different semantics" is my least favourite option, because it means if you transition from HTTP+BiDi to BiDi-only then things mysteriously break without even an error message. For example if you were sending `"dismiss"` as the default prompt handler then suddenly any navigation with a `beforeUnload` prompt will be silently cancelled. Subsetting the allowed values at least makes it clear as soon as you start the session what the problem is.

Reply:
Note: as discussed in https://github.com/w3c/webdriver-bidi/pull/681#issuecomment-2163474276, the changing `default` semantics is not a concern from the user side. So we decided to go with: "BiDi `default` covers all the prompts".